lib/Internal/Windows/SocketConnector.php

```diff
@@ -0,0 +1,326 @@
+    {
+        \fwrite($socket, \chr(SignalCode::HANDSHAKE_ACK) . \chr($code));
+        \fclose($socket);
+        unset($this->pendingClients[(int)$socket]);
+    }
+
+    private function failHandleStart(Handle $handle, string $message, ...$args)
+    {
+        Loop::cancel($handle->connectTimeoutWatcher);
+        unset($this->pendingProcesses[$handle->wrapperPid]);
+
+        foreach ($handle->sockets as $socket) {
+            \fclose($socket);
+        }
+
+        $handle->startDeferred->fail(new ProcessException(\vsprintf($message, $args)));
+    }
+
+    /**
+     * Read data from a client socket
+     *
+     * This method cleans up internal state as appropriate. Returns null if the read fails or needs to be repeated.
+     *
+     * @param resource $socket
+     * @param int $length
+     * @param PendingSocketClient $state
+     * @return string|null
+     */
+    private function readDataFromPendingClient($socket, int $length, PendingSocketClient $state)
+    {
+        $data = \fread($socket, $length);
+
+        if ($data === false || $data === '') {
+            \fclose($socket);
+            Loop::cancel($state->readWatcher);
+            Loop::cancel($state->timeoutWatcher);
+            unset($this->pendingClients[(int)$socket]);
+            return null;
+        }
+
+        $data = $state->recievedDataBuffer . $data;
+
+        if (\strlen($data) < $length) {
+            $state->recievedDataBuffer = $data;
+            return null;
+        }
+
+        $state->recievedDataBuffer = '';
+
+        Loop::cancel($state->readWatcher);
+        Loop::cancel($state->timeoutWatcher);
+
+        return $data;
+    }
+
+    public function onReadable_Handshake($watcher, $socket) {
+        $socketId = (int)$socket;
+        $pendingClient = $this->pendingClients[$socketId];
+
+        if (null === $data = $this->readDataFromPendingClient($socket, self::SECURITY_TOKEN_SIZE + 6, $pendingClient)) {
+            return;
+        }
+
+        $packet = \unpack('Csignal/Npid/Cstream_id/a*client_token', $data);
+
+        // validate the client's handshake
+        if ($packet['signal'] !== SignalCode::HANDSHAKE) {
+            $this->failClientHandshake($socket, HandshakeStatus::SIGNAL_UNEXPECTED);
+            return;
+        }
+
+        if ($packet['stream_id'] > 2) {
+            $this->failClientHandshake($socket, HandshakeStatus::INVALID_STREAM_ID);
+            return;
+        }
+
+        if (!isset($this->pendingProcesses[$packet['pid']])) {
+            $this->failClientHandshake($socket, HandshakeStatus::INVALID_PROCESS_ID);
+            return;
+        }
+
+        $handle = $this->pendingProcesses[$packet['pid']];
+
+        if (isset($handle->sockets[$packet['stream_id']])) {
+            $this->failClientHandshake($socket, HandshakeStatus::DUPLICATE_STREAM_ID);
+            $this->failHandleStart($handle, "Received duplicate socket for stream #%d", $packet['stream_id']);
```

Comment:
The PID is something sent by the connecting client and nothing secret. The security token only exists to protect against suspicious behavior of other processes.